Core/Object Arts/Dolphin/MVP/Views/Common Controls/ListView.cls

```diff
@@ -1474,41 +1474,51 @@ moreParamsAt: n put: anObject
 	^self moreParams at: n put: anObject!
 
 newSelectionsFromEvent: event
-	"Private - Answer the <integer> selection that would occur if the <MouseEvent>,
+	"Private - Answer a collection of the <integer> selections that would occur if the <MouseEvent>,
 	event, was was passed to the control."
 
 	| itemClicked anyKeysDown newSelections markIndex markRange markIsSelected |
 	itemClicked := self itemFromPoint: event position.
+	"In single-select mode, the new selection is always the item clicked, or nothing if the click is outside all items."
 	self isMultiSelect ifFalse: [^itemClicked ifNil: [#()] ifNotNil: [:item | Array with: item]].
 	newSelections := self selectionsByIndex.
-	""
 	anyKeysDown := event isShiftDown or: [event isCtrlDown].
-	((itemClicked isNil and: [anyKeysDown])
-		or: [event button ~= #left and: [anyKeysDown or: [newSelections includes: itemClicked]]])
-			ifTrue: [^newSelections].
-	""
+	"If the click is outside all items, and ctrl or shift is down, nothing changes."
+	(itemClicked isNil and: [anyKeysDown]) ifTrue: [^newSelections].
+	"A right or middle click on an already-selected item also does not change the selection,
+	otherwise it would be very frustrating to open the context menu with multiple items selected.
+	The control also chooses to leave the selection alone if modifiers are held during a right/middle click."
+	(event button ~= #left and: [anyKeysDown or: [newSelections includes: itemClicked]])
+		ifTrue: [^newSelections].
+	"Having taken care of the above edge cases, if no keys are down, the behavior is equivalent to the single-select case."
 	anyKeysDown ifFalse: [^itemClicked ifNil: [#()] ifNotNil: [:item | Array with: item]].
-	"Only Ctrl"
+	"Only Ctrl is down, so we flip the selection state of the clicked item."
 	event isShiftDown ifFalse: 
 			[newSelections := newSelections asSortedCollection.
```

Comment:
Not code you are changing here, but is it ever the case that the `selectionsByIndex` returns an unordered list? I don't think so as `ListView>>getMultipleSelections` says the selections will be in ascending order. Sorting an already ordered list is a pathologically slow case for quicksort, although probably this wouldn't be an issue in practice as Dolphin uses a modified quicksort by default, and the number of selected items is not likely to be very large. A good reason to use a `SortedCollection` would be to take advantage of an optimised `remove:ifAbsent:`, e.g. using a binary search, but currently the implementation isn't optimised and will perform as for `OrderedCollection`.